woocommerce/woocommerce-ios
# Login - Jetpack Setup: Add tracks events

Part of: #8075 


### Description
This PR adds tracking events for the native Jetpack installation project.

I apologize for the 500+ lines of change. It is due to unit tests. 🙏 

### Testing instructions

Prerequisite
1. You need site A, which doesn't have Jetpack installed.
1. You need site B, which has Jetpack installed but isn't connected to your WPCOM account. 


Steps
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- If your account doesn't have any associated site, notice that the empty store picker is displayed with the matching design as on Figma.
- Select Add a Store > Connect an existing site. (Either site A or site B from the Prerequisite)
- Enter the address of your test store and tap Continue. 
  - Upon successful site discovery you should see a event like the following in Xcode log.
    ```
    Tracked site_picker_site_discovery, properties: [AnyHashable("has_wordpress"): true, AnyHashable("is_wpcom"): false, AnyHashable("is_jetpack_connected"): false, AnyHashable("is_jetpack_active"): false, AnyHashable("is_jetpack_installed"): false]
    ```
- You will land on the Jetpack error screen.
  - If you have entered Site A (No Jetpack installed)
    - Check Xcode log and ensure that `login_jetpack_required_screen_viewed` event is logged
    - Tap on "Install Jetpack" button and ensure that `login_jetpack_setup_button_tapped` is logged
  - If you entered Site B (Only Jetpack connection missing)
  - Check Xcode log and ensure that `login_jetpack_connection_error_shown` event is logged
  - Tap on "Install Jetpack" button and ensure that `login_jetpack_connect_button_tapped` is logged
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- You will hand on the site credential screen
  - Ensure `login_jetpack_site_credential_screen_viewed` event is logged.
  - Dismiss the screen and ensure `login_jetpack_site_credential_screen_dismissed` event is logged.
  - Tap "Reset password" and ensure `login_jetpack_site_credential_reset_password_button_tapped` is logged.
  - Enter invalid creds and tap "Install Jetpack" button and notice the error alert. Check logs and ensure `login_jetpack_site_credential_did_show_error_alert` is logged.
  - Tapping "Install Jetpack" button should log `login_jetpack_site_credential_install_button_tapped`
  - Enter the correct credentials for your store and `login_jetpack_site_credential_did_finish_login` should be logged.
- You will land on the Jetpack setup screen. 
  - `login_jetpack_setup_screen_viewed` will logged.
  - Dismiss the screen and ensure `login_jetpack_setup_screen_dismissed` is logged.
  - We have events for each step's success/failure. You will have to mock the responses using a proxy tool to test the error events properly. But I think testing the success events is enough. 
  - For successful installation and activation you should see the following events.
  - `login_jetpack_install_install_successful` -> `login_jetpack_install_activation_successful` -> `login_jetpack_install_fetch_jetpack_connection_url_successful` -> `login_jetpack_install_all_steps_marked_done`
  - When the setup reaches the connection step, notice that a web view is presented for approving Jetpack connection.
  - Tap on the Authorize button on the web page, notice that the web view is then dismissed, and the Jetpack setup screens should then show that all steps are completed.
  - Tap on "Go to Store" and ensure that `login_jetpack_install_go_to_store_button_tapped` is logged.

### Screenshots

**Jetpack error screen**
| Without Jetpack | With Jetpack installed but no WPCOM account connection |
| - | - |
| ![Simulator Screen Shot - iPhone 11 Pro - 2022-11-22 at 12 04 48](https://user-images.githubusercontent.com/524475/203505207-e4ebc737-d77e-4f9a-9325-d4a856aba1c9.png) | ![Simulator Screen Shot - iPhone 11 Pro - 2022-11-22 at 12 03 44](https://user-images.githubusercontent.com/524475/203505195-fdbb5548-c203-4745-a175-442e256b482f.png) |

**Site credentials screen**
| Normal | With error alert |
| - | - |
| ![Simulator Screen Shot - iPhone 11 Pro - 2022-11-22 at 12 36 30](https://user-images.githubusercontent.com/524475/203505698-4583ca85-ce81-4014-9d3d-bb9060d93d51.png) | ![Simulator Screen Shot - iPhone 11 Pro - 2022-11-22 at 12 36 00](https://user-images.githubusercontent.com/524475/203505679-7e997a97-6327-455f-bc7d-6f247aca7ee2.png) |

**Jetpack setup screen**
| Installation in progress | Installation complete |
| - | - |
| ![Simulator Screen Shot - iPhone 11 Pro - 2022-11-22 at 12 57 10](https://user-images.githubusercontent.com/524475/203505909-0bbb4906-9f83-4324-b9ae-4a3c4fb3a5d4.png) | ![Simulator Screen Shot - iPhone 11 Pro - 2022-11-22 at 13 59 45](https://user-images.githubusercontent.com/524475/203505925-94288604-3008-4bce-a330-8e6fc1d96bf5.png) |

---
